VariantShredding.md

```diff
@@ -25,290 +25,316 @@
+The `typed_value` field may be omitted when not shredding elements as a specific type.
+When `typed_value` is omitted, `value` must be `required`.
 
-1) In a case where there are rare type mismatches (for example, a numeric field with rare strings like “n/a”), we allow the field to be shredded, which could still be a significant performance benefit compared to fetching and decoding the full value/metadata binary.
-2) Since there is a single schema per file, there would be no easy way to recover from a type mismatch encountered late in a file write. Parquet files can be large, and buffering all file data before starting to write could be expensive. Including a variant column for every field guarantees we can adhere to the requested shredding schema.
+For example, a `tags` Variant may be shredded as a list of strings using the following definition:
+```
+optional group tags (VARIANT) {
+  required binary metadata;
+  optional binary value;
+  optional group typed_value (LIST) {   # must be optional to allow a null list
+    repeated group list {
+      required group element {          # shredded element
+        optional binary value;
+        optional binary typed_value (STRING);
+      }
+    }
+  }
+}
+```
 
-# Top-level metadata
+All elements of an array must be present (not missing) because the `array` Variant encoding does not allow missing elements.
+That is, either `typed_value` or `value` (but not both) must be non-null.
+Null elements must be encoded in `value` as Variant null: basic type 0 (primitive) and physical type 0 (null).
 
-Any values stored in a shredded `variant_value` field may have dictionary IDs referring to the metadata.
-There is one metadata value for the entire Variant record, and that is stored in the top-level `metadata` field.
-This means any `variant_value` values in the shredded representation is only the "value" portion of the [Variant Binary Encoding](VariantEncoding.md).
+The series of `tags` arrays `["comedy", "drama"], ["horror", null], ["comedy", "drama", "romance"], null` would be stored as:
 
-The metadata is kept at the top-level, instead of shredding the metadata with the shredded variant values because:
-* Simplified shredding scheme and specification. No need for additional struct-of-binary values, or custom concatenated binary scheme for `variant_value`.
-* Simplified and good performance for write shredding. No need to rebuild the metadata, or re-encode IDs for `variant_value`.
-* Simplified and good performance for Variant reconstruction. No need to re-encode IDs for `variant_value`.
+| Array                            | `value`     | `typed_value `| `typed_value...value` | `typed_value...typed_value`    |
+|----------------------------------|-------------|---------------|-----------------------|--------------------------------|
+| `["comedy", "drama"]`            | null        | non-null      | [null, null]          | [`comedy`, `drama`]            |
```

Comment:
it might be worth describing def/repetition level here tobe more explicit?

Reply:
I don't think that would help make this table more clear. All of those are `optional`. The only case where `value` is `required` is when `typed_value` is omitted, which is not the case.